### ISSUE TITLE
Project's info request

### ISSUE THREAD

**Issue by shegeley** [guile-wayland/guile-wayland#1]
Hi. Really cool project. 👍🏽 

I would also like to have guile bindings for wayland.
I see the project is WIP for more than a year. I guess it's hard to wrap all the wayland calls into scheme.  It's also the most complete opensource project in the direction of wayland+guile I could find. 

I would like to see more of the projects's background and maybe some examples. 

From guix.scm I see that you are using custom bytestructures from the directory upper that the project in yours file system. 

```scheme
(primitive-load
       (string-append (dirname (dirname (current-filename)))
                      "/guile-bytestructure-class/guix.scm"))
```
So it can't be just copy-pasted and easily installed on guix. 😞 

It's just a big and monolithic task: to pack all the wayland in one go. Maybe there are possibility to provide some minimal wrapping and make it extendable?

**Comment by Z572** [comment]
Hello, this project mainly serves the guile-wlroots project. Currently, only part of the wayland-server ffi package is provided. It may not be possible to use this project alone to run a window manager. If you want to contribute, you are welcome. Also maybe you can take a look at my gwwm project. and maybe you can take a look at https://github.com/mwette/guile-wl-play, but I haven't tried it (use Google translation)

**Comment by shegeley** [comment]
Thanks. Sorry for a late reply. Let me rephrase my request. 

I see wayland (implementations) as a pretty complex software: it constantly grows and has new protocols added. I'd like to get into calling it explicitly with Guile Scheme.  As you have some experience with it, please suggest how'd it better be structured in your opinion.

Although I only want (for now) to call a few unstable protocols from Wayland, I'm thinking it terms on composable code: how would I be able to reuse it best and give others this ability. 

I guess it's a bad idea to have one huge "official" `guile-wayland` package (piece of software) that would "include it all".  Something like: have a "core" package with bytestructure packaging of every FFI C call response and then have all the protocols have logical namespacing (module-naming) like `(wayland fullscreen-shell)` and etc? [edited: append] that might be distributed as a separate packages having only core dependency that declares some FFI byte-structure conventions

**Comment by Z572** [comment]
some bind have a `wayland-scanner` program to convents xml to code, I think should add a "guile's language" in the future, such as `guile-xcb` transform xml file to .go file ,but for now just of write every module by people and [a simple script](https://github.com/Z572/guile-wayland/blob/master/build-aux/wayland-scanner-guile.in) gen some enum to `(wayland protocol)`,

 (wayland protocols fullscreen-shell server)
in the future
guile-wayland support's protocols are wrap, modify and re-export from `(wayland protocols fullscreen-shell server)` to `(wayland fullscreen-shell server)`, and provide a easy use use interface,

unsupported protocols  just like guile-gi have a `use-typelibs` macro can import GObject Introspection,  have a `(use-server-protocol "any-protocol")` compile "any-protocol" to a `(wayland protocols any-protocol server)` module

**Comment by shegeley** [comment]
I'm thinking about how it'd be possible to have some super-minimal core package that'd include anything you'd need to write your own protocol lib with guile. So, that one person/repository wouldn't need to be responsible for all huge wayland protocols implementation.

**Comment by Z572** [comment]
just wrap wayland-util.h, wayland-client-core.h,wayland-server-core.h and wayland-cursor.h, and write a program or macro to transform xml file, other protocol lib should can base on this.

**Comment by shegeley** [comment]
thanks 👍
